pydicom/dataelem.py

```diff
@@ -535,8 +546,14 @@ def __eq__(self, other):
             return True
 
         if isinstance(other, self.__class__):
-            return (self.tag == other.tag and self.VR == other.VR
-                    and self.value == other.value)
+            if self.tag == other.tag and self.VR == other.VR:
+                if have_numpy and isinstance(self.value, numpy.ndarray):
+                    return (len(self.value) == len(other.value)
+                            and numpy.allclose(self.value, other.value))
+                else:
+                    return self.value == other.value
+            else:
```

Comment:
I think both are needed.  One is the case where tag or VR doesn't match, the other is for the value being numpy or not.  But I will rearrange to a clearer order by negating the tag, VR check and returning False right away.

Reply:
I see you already got rid of the first one. I meant you don't need an `else` after a return:
```python
            if config.have_numpy and isinstance(self.value, numpy.ndarray):
                if (len(self.value) == len(other.value)
                        and numpy.allclose(self.value, other.value)
                return (len(self.value) == len(other.value)
                        and numpy.allclose(self.value, other.value))
            return self.value == other.value
        return NotImplemented
```
(don't change it, this is only to explain what I meant)

Reply:
Ah, I see.  Yeah, I usually try to avoid `return` inside `if` blocks, especially larger ones, because it can be harder to follow the logic later, so in this case I didn't think of it.  The latest commit makes the logic a lot clearer anyway I think, with a very simple return.